tsdb/index/labelvalues_test.go

```diff
@@ -0,0 +1,195 @@
+		labels.FromStrings("a", "1", "b", "4"),
+		labels.FromStrings("a", "2", "b", "5"),
+	}
+
+	require.NoError(t, iw.AddSymbol("1"))
+	require.NoError(t, iw.AddSymbol("2"))
+	require.NoError(t, iw.AddSymbol("3"))
+	require.NoError(t, iw.AddSymbol("4"))
+	require.NoError(t, iw.AddSymbol("5"))
+	require.NoError(t, iw.AddSymbol("a"))
+	require.NoError(t, iw.AddSymbol("b"))
+
+	// Postings lists are only written if a series with the respective
+	// reference was added before.
+	require.NoError(t, iw.AddSeries(1, series[0]))
+	require.NoError(t, iw.AddSeries(2, series[1]))
+	require.NoError(t, iw.AddSeries(3, series[2]))
+	require.NoError(t, iw.AddSeries(4, series[3]))
+	require.NoError(t, iw.AddSeries(5, series[4]))
+
+	require.NoError(t, iw.Close())
+
+	ir, err := NewFileReader(fn)
+	require.NoError(t, err)
+
+	// Obtain postings where a=1
+	p, err := ir.Postings(ctx, "a", "1")
+	require.NoError(t, err)
+
+	it := ir.LabelValuesExcluding(p, "b")
+	var vals []string
+	for it.Next() {
+		vals = append(vals, it.At())
+	}
+	require.NoError(t, it.Err())
+	require.Empty(t, it.Warnings())
+	// Note that only b=5 is not filtered out, since it has a=2, while our postings have a=1
+	require.Equal(t, []string{"5"}, vals)
+
+	t.Run("empty result set", func(t *testing.T) {
+		it := ir.LabelValuesExcluding(p, "c")
+		require.False(t, it.Next())
+		require.NoError(t, it.Err())
+		require.Empty(t, it.Warnings())
+	})
+}
+
+func TestMemPostings_LabelValuesFor(t *testing.T) {
+	mp := NewMemPostings()
+	mp.Add(1, labels.FromStrings("a", "1"))
+	mp.Add(1, labels.FromStrings("b", "1"))
+	mp.Add(2, labels.FromStrings("a", "1"))
+	mp.Add(2, labels.FromStrings("b", "2"))
+	mp.Add(3, labels.FromStrings("a", "1"))
+	mp.Add(3, labels.FromStrings("b", "3"))
+	mp.Add(4, labels.FromStrings("a", "1"))
+	mp.Add(4, labels.FromStrings("b", "4"))
+	mp.Add(5, labels.FromStrings("a", "2"))
+	mp.Add(5, labels.FromStrings("b", "5"))
+	p := mp.Get("a", "1")
+
+	it := mp.LabelValuesFor(p, "b")
+
+	var vals []string
+	for it.Next() {
+		vals = append(vals, it.At())
+	}
+	require.NoError(t, it.Err())
+	require.Empty(t, it.Warnings())
+	// Note that b=5 is filtered out, since it has a=2, while our postings have a=1
+	require.Equal(t, []string{"1", "2", "3", "4"}, vals)
+
+	t.Run("empty result set", func(t *testing.T) {
+		it := mp.LabelValuesFor(p, "c")
+		require.False(t, it.Next())
+		require.NoError(t, it.Err())
+		require.Empty(t, it.Warnings())
+	})
+}
+
+func TestMemPostings_LabelValuesExcluding(t *testing.T) {
+	mp := NewMemPostings()
+	mp.Add(1, labels.FromStrings("a", "1"))
+	mp.Add(1, labels.FromStrings("b", "1"))
+	mp.Add(2, labels.FromStrings("a", "1"))
+	mp.Add(2, labels.FromStrings("b", "2"))
+	mp.Add(3, labels.FromStrings("a", "1"))
+	mp.Add(3, labels.FromStrings("b", "3"))
+	mp.Add(4, labels.FromStrings("a", "1"))
+	mp.Add(4, labels.FromStrings("b", "4"))
+	mp.Add(5, labels.FromStrings("a", "2"))
+	mp.Add(5, labels.FromStrings("b", "5"))
```

Comment:
We're adding a labelset for each series, not appending individual labels to series (that's not a valid operation, and it only casually works because of current implementation of the method).

```suggestion
	mp.Add(1, labels.FromStrings("a", "1", "b", "1"))
	mp.Add(2, labels.FromStrings("a", "1", "b", "2"))
	mp.Add(3, labels.FromStrings("a", "1", "b", "3"))
	mp.Add(4, labels.FromStrings("a", "1", "b", "4"))
	mp.Add(5, labels.FromStrings("a", "2", "b", "5"))
```

Can you also review the rest of the test cases please?